File: examples/Python/MR/acquisitions_handling.py
'''
Upper-level interface demo that illustrates how MR data can be interfaced 
from python.

Usage:
  acquisitions_handling.py [--help | options]

Options:
  -f <file>, --file=<file>    raw data file
                              [default: simulated_MR_2D_cartesian.h5]
  -p <path>, --path=<path>    path to data files, defaults to data/examples/MR
                              subfolder of SIRF root folder
  -r <rnge>, --range=<rnge>   range of acquisitions to examine as string '(a,b)'
                              [default: (254, 258)]
  -e <engn>, --engine=<engn>  reconstruction engine [default: Gadgetron]
'''

__version__ = '0.1.0'
from docopt import docopt
args = docopt(__doc__, version=__version__)

from ast import literal_eval
import os
import sys

# import engine module
exec('from p' + args['--engine'] + ' import *')

def main():

    # locate the input data file
    data_path = args['--path']
    if data_path is None:
        data_path = mr_data_path()
    input_file = existing_filepath(data_path, args['--file'])

    # acquisitions will be read from an HDF file input_file
    input_data = AcquisitionData(input_file)

    # Get number of acquisitions:
    # the raw k-space data is a list of different 1D 
    # acquisitions (readouts) of different data type (e.g. noise correlation 
    # data, navigator data, image data,...).
    # The number of all aquisitions is
    na = input_data.number_of_acquisitions()
    # The number of image data acquisitions is
    ni = input_data.number_of_acquisitions('image')
    print('acquisitions: total %d, image data %d' % (na, ni))

    # sort data acquisition
    input_data.sort()

    # retrieve the range of acquisitions to examine
    t = literal_eval(args['--range'])
    if t[0] >= t[1] or t[1] >= na:
        raise error('Wrong acquisitions range')
    where = range(t[0], t[1])

    # retrieve acquisitions flags
    flags = input_data.get_info('flags')

    # inspect the first acquisition flag
    if flags[0] & IMAGE_DATA_MASK:
        print('first acquisition is image data')
    else:
        # should see this if input data file is test_2D_2x.h5
        print('first acquisition is not image data')
        
    # display flags
    print('Flags'),
    print(flags[where])
    
    # inspect some kspace_encode_step_1 counters
    es1 = input_data.get_info('encode_step_1')
    print('Ky/PE - encoding'),
    print(es1[where])
    
    # inspect some slice counters
    slices = input_data.get_info('slice')
    print('Slices'),
    print(slices[where])
    
    # inspect some repetition counters
    print('Repetitions'),
    reps = input_data.get_info('repetition')
    print(reps[where])

    # copy raw data into python array and determine its size
    # in the case of the provided dataset 'simulated_MR_2D_cartesian.h5' the 
    # size is 2x256 phase encoding, 8 receiver coils and points 512 readout 
    # points (frequency encoding dimension)
    input_array = input_data.as_array()
    input_shape = input_array.shape
    print('input data dimensions: %dx%dx%d' % input_shape)

    # pre-process acquisitions
    print('---\n pre-processing acquisitions...')
    processed_data = preprocess_acquisitions(input_data)

    # copy processed acquisitions into an array and determine its size
    # by removing the oversampling factor of 2 along the readout direction, the
    # number of readout samples was halfed
    processed_array = processed_data.as_array()
    processed_shape = processed_array.shape
    print('processed data dimensions: %dx%dx%d' % processed_shape)

try:
    main()
    print('done')

except error as err:
    # display error information
    print('??? %s' % err.value)

File: examples/Matlab/MR/acquisitions_handling.m
function acquisitions_handling(engine)
% ACQUISITIONS_HANDLING Demo illustrating acquisitions pre-processing 
% and plotting.
%
% In MATLAB, there are also ISMRMRD tools available for examining 
% data before processing.
%


% Select and import SIRF MATLAB MR package so that SIRF MR objects can be 
% created in this function without using the prefix 'MR.'
%set_up_mr
%import MR.*
if nargin < 1
    engine = [];
end
eval(setup_MR(engine))

% acquisitions will be read from an HDF file
[filename, pathname] = uigetfile('*.h5', 'Select raw data file', mr_data_path);
input_data = AcquisitionData(fullfile(pathname, filename));

na = input_data.number();
fprintf('%d acquisitions (readouts) found\n', na)

fprintf('sorting acquisitions...\n')
input_data.sort()

% dimensions of k-space for one slice?
[ns, ny, nc] = input_data.slice_dimensions();

% dimensions method returns size of all (i.e. including noise data) if 
% argument is passed in or if 'all' is passed in. Passing in anything else
% means not all !!
[ns, nc, na] = input_data.dimensions('not all');

% pre-process acquisition data
fprintf('processing acquisitions...\n')
processed_data = preprocess_acquisitions(input_data);
processed_data.sort()

% selected methods for getting information
flags = input_data.get_info('flags');
es1 = input_data.get_info('encode_step_1');
slice = input_data.get_info('slice');
repetition = input_data.get_info('repetition');

while true
    num = input('enter acquisition number (0 to stop this loop): ');
    if num < 1 || num > na
        break
    end
    a = input_data.acquisition(num); % There are methods for getting selected 
                                     % properties such as flags, but not
                                     % the complete set (user may be better
                                     % using MATLAB ISMRMRD tools)
    fprintf('samples: %d\n', a.number_of_samples())
    fprintf('flags: %d %d\n', a.flags(), flags(num))
    fprintf('encode step 1: %d %d\n', a.idx_kspace_encode_step_1(), es1(num))
    fprintf('slice: %d %d\n', a.idx_slice(), slice(num))
    fprintf('repetition: %d %d\n', a.idx_repetition(), repetition(num))
end

% Data returned as complex array
iarr0 = input_data.as_array();
parr0 = processed_data.as_array();

is = ns/2;
ic = nc/2;
ia = na/2;
disp(['Value of one array element: '])
disp(iarr0(is, ic, ia))

iarr0(is, ic, ia) = iarr0(is, ic, ia)*10;

% Data can be replaced using fill method
input_data.fill(iarr0);
processed_data.fill(parr0);

iarr = input_data.as_array();
parr = processed_data.as_array();

disp(['Value of same array element after replacement with 10x data: '])
disp(iarr(is, ic, ia))

iarr = abs(iarr);
parr = abs(parr);
iarr = iarr/max(max(max(iarr)));
parr = parr/max(max(max(parr)));

nz = idivide(na, ny);

has_ipt = exist('imadjust','file') && exist('mat2gray','file');

fprintf...
    ('Enter z-coordinate of the slice to view the acquired data for it\n')
fprintf...
    ('(a value outside the range [1 : %d] will stop this loop)\n', nz)
while (true)
    z = int32(input('slice: '));
    if z < 1 || z > nz
        break
    end
    fprintf('Enter coil number to view the acquired data for it\n')
    fprintf...
        ('(a value outside the range [1 : %d] will stop this loop)\n', nc)
    while (true)
        c = input('coil: ');
        if c < 1 || c > nc
            break
        end
        figure('Name',['Input Data as array, coil: ',num2str(c)])
        data = squeeze(iarr(:, c, ny*(z - 1) + 1 : ny*z));
        if has_ipt
            imshow(imadjust(mat2gray(abs(data)),[0 0.7],[],0.2))
        else
            imshow(data)
        end
        title('Input data')
        cc = double(c + nc);
        figure('Name',['Processed data as array, coil: ',num2str(c)])
        data = squeeze(parr(:, c, ny*(z - 1) + 1 : ny*z));
        if has_ipt
            imshow(imadjust(mat2gray(abs(data)),[0 0.7],[],0.2))
        else
            imshow(data)
        end
        title('Processed data')
    end
end


